src/databricks/labs/ucx/source_code/graph.py

```diff
@@ -589,6 +588,37 @@ def finalize(self) -> InheritedContext:
         return InheritedContext(tree, self.found)
 
 
+class Lineage(abc.ABC):
+
+    @abc.abstractmethod
+    def to_json(self) -> Any: ...
+
+
+class PathLineage(Lineage):
+
+    def __init__(self, path: Path):
+        self._path = path
+
+    def to_json(self) -> Any:
+        return str(self._path)
+
+
+class CompositeLineage(Lineage):
+
+    def __init__(self, *lineages: Lineage):
+        self._lineages = lineages
+
+    def to_json(self) -> Any:
+        atoms: list[Any] = []
```

Comment:
this is not deterministic enough. we have to guarantee serialization to a sequence of tuples at `mypy` time - e.g. `[["path", "/foo/bar"], ["task", "some"], ["job", "923789172982"]]` (or the reverse one).